node/node_newblock.go

```diff
@@ -40,28 +43,40 @@ func (node *Node) WaitForConsensusReadyV2(readySignal chan struct{}, stopChan ch
 				utils.Logger().Debug().
 					Msg("Consensus new block proposal: STOPPED!")
 				return
-			case <-readySignal:
-				for node.Consensus != nil && node.Consensus.IsLeader() {
+			case proposalType := <-readySignal:
+				if node.Consensus != nil && node.Consensus.IsLeader() {
 					time.Sleep(SleepPeriod)
 					utils.Logger().Info().
 						Uint64("blockNum", node.Blockchain().CurrentBlock().NumberU64()+1).
 						Msg("PROPOSING NEW BLOCK ------------------------------------------------")
 
 					// Prepare last commit signatures
-					commitSigs := make(chan []byte)
-					sigs, err := node.Consensus.BlockCommitSigs(node.Blockchain().CurrentBlock().NumberU64())
+					newCommitSigsChan := make(chan []byte)
 
-					node.Consensus.StartFinalityCount()
-					if err != nil {
-						utils.Logger().Error().Err(err).Msg("[ProposeNewBlock] Cannot get commit signatures from last block")
-						break
-					}
-					// Currently the block proposal is not triggered asynchronously yet with last consensus.
-					// TODO: trigger block proposal when 66% commit, and feed and final commit sigs here.
 					go func() {
-						commitSigs <- sigs
+						waitTime := 0 * time.Second
+						if proposalType == consensus.AsyncProposal {
+							waitTime = 4 * time.Second
```

Comment:
why 4 seconds here?

Reply:
4 seconds is just a number that's large enough for timeout waiting for 100% sig. usually it takes 1s for the committed sig to be collected fully. And our grace period is also less than 4 seconds.